mission_control/navigator_missions/navigator_missions/entrance_gate.py

```diff
@@ -0,0 +1,463 @@
+                                   pinger_vector[0] - gate_vector[0])
+            intersection = (1 - t) * gate_vector[0] + t * gate_vector[1]
+
+            # Calculate the distances from each gate center to the intersection
+            gateDist[0] = np.linalg.norm(intersection - np.array(self.gate_centers[0]))
+            gateDist[1] = np.linalg.norm(intersection - np.array(self.gate_centers[1]))
+            gateDist[2] = np.linalg.norm(intersection - np.array(self.gate_centers[2]))
+
+            # If the intersection is more than 15 meters from the center of the gates it is outside the gates
+            if gateDist[1] > 16:
+                # Ignore it
+                continue
+
+            # Calculate vectors to the edges of the gates and the intersect
+            vec_corner1 = gate_vector[0] - pinger_vector[0]
+            vec_corner1 = vec_corner1 / np.linalg.norm(vec_corner1)
+            vec_corner4 = gate_vector[1] - pinger_vector[0]
+            vec_corner4 = vec_corner4 / np.linalg.norm(vec_corner4)
+            vec_intersect = intersection - pinger_vector[0]
+            vec_intersect = vec_intersect / np.linalg.norm(vec_intersect)
+
+            # Calculate the angle between the edges of the gates and the angle between an edge and the intersection
+            ang_c1_c4 = np.arccos(np.clip(np.dot(vec_corner1, vec_corner4), -1.0, 1.0))
+            ang_c1_i4 = np.arccos(np.clip(np.dot(vec_corner1, vec_corner4), -1.0, 1.0))
+
+            # If the angles are out of range, throw out this intersection
+            if ang_c1_c4 >= 0 and (ang_c1_i4 < 0 or ang_c1_i4 > ang_c1_c4):
+                continue
+            if ang_c1_c4 < 0 and (ang_c1_i4 > 0 or ang_c1_i4 < ang_c1_c4):
+                continue
+
+            # Determine which gate the intersection is in
+            bucket_index = np.argmin(np.array(gateDist))
+            bucket[bucket_index] = bucket[bucket_index] + 1
+
+        # Determine which gate has the most hits
+        pinger_gate = np.argmax(np.array(bucket))
+
+        # Calculate the confidence
+        confidence = bucket[pinger_gate] / (bucket[0] + bucket[1] + bucket[2])
+        return (pinger_gate, confidence)
+
+    @util.cancellableInlineCallbacks
+    def ping_recv(self, p_message):
+        try:
+            # Transform the ping into enu
+            hydrophones_to_enu = yield self.tf_listener.get_transform('enu', p_message.header.frame_id)
+            hydrophones_origin = hydrophones_to_enu._p[0:2]
+            heading = rosmsg_to_numpy(p_message.vector)
+            heading_enu = hydrophones_to_enu.transform_vector(heading)
+            heading_enu = heading_enu[0:2] / np.linalg.norm(heading_enu[0:2])
+
+            # Track the ping
+            self.intersect_vectors.append((hydrophones_origin[0:2], hydrophones_origin[0:2] + heading_enu[0:2]))
+        except tf2_ros.TransformException, e:
+            rospy.logwarn('TF Exception: {}'.format(e))
```

Comment:
can't rospy in txros program

Reply:
Fixed https://github.com/jaxnb/NaviGator/commit/e26de47ddb4335edd1105f6102eb5d21daf7e730